### ISSUE TITLE
URLs containing special characters are escaped

### ISSUE THREAD

**Issue by Neamar** [linkchecker/linkchecker#203]
## Summary
The following URL works in a browser: [https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_?](https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_?)

However, it will fail in linkchecker

## Steps to reproduce

 `linkchecker https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_\?`

(reverse slash included to quote in bash, not part of the URL)

## Actual result
```
URL        `https://omnilogie.fr/O/Doù_vient_la_Légion_dHonneur_?'
Real URL   https://omnilogie.fr/O/D'o%F9_vient_la_L%E9gion_d'Honneur_
Check time 0.886 seconds
Size       0B
Info       Redirected to
           `https://omnilogie.fr/O/D'o%F9_vient_la_L%E9gion_d'Honneur_'.
Result     Error: TooManyRedirects: Exceeded 10 redirects.
```
## Expected result
URL is crawled correctly

## Environment

 * Operating system: Ubuntu 18.04.1 LTS
 * Linkchecker version: `LinkChecker 9.4.0`
 * Python version: Python 2.7.15rc1
 * Install method: `sudo pip install git+https://github.com/linkchecker/linkchecker.git`
 * Site URL: https://omnilogie.fr

## Logs

```
linkchecker --verbose https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_\?
INFO linkcheck.cmdline 2019-01-14 23:37:46,136 MainThread Checking intern URLs only; use --check-extern to check extern URLs.
LinkChecker 9.4.0              Copyright (C) 2000-2014 Bastian Kleineidam
LinkChecker comes with ABSOLUTELY NO WARRANTY!
This is free software, and you are welcome to redistribute it
under certain conditions. Look at the file `LICENSE' within this
distribution.
Get the newest version at http://wummel.github.io/linkchecker/
Write comments and bugs to https://github.com/linkchecker/linkchecker/issues
Support this project at http://wummel.github.io/linkchecker/donations.html

Start checking at 2019-01-14 23:37:46+001

URL        `https://omnilogie.fr/O/Doù_vient_la_Légion_dHonneur_?'
Real URL   https://omnilogie.fr/O/D'o%F9_vient_la_L%E9gion_d'Honneur_
Check time 0.754 seconds
Size       0B
Info       Redirected to
           `https://omnilogie.fr/O/D'o%F9_vient_la_L%E9gion_d'Honneur_'.
Result     Error: TooManyRedirects: Exceeded 10 redirects.

Statistics:
Downloaded: 0B.
Content types: 0 image, 1 text, 0 video, 0 audio, 0 application, 0 mail and 0 other.
URL lengths: min=58, max=58, avg=58.

That's it. 1 link in 2 URLs checked. 0 warnings found. 1 error found.
Stopped checking at 2019-01-14 23:37:46+001 (0.77 seconds)

```

## Other notes
A curl on the same URL succeeds: ` curl -v "https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_?"`.

I think the issue is probably caused by two different things:

* special characters in the URL (accents)
* the fact that the URL uses a question mark, which is valid but not super common as it usually indicate the start of GET parameters

**Comment by anarcat** [comment]
the steps to reproduce are incorrect:

    linkchecker https://omnilogie.fr/O/D'où_vient_la_Légion_d'Honneur_\?

will fail, because `'` is special to most shells. however, I can reproduce with:

    linkchecker "https://omnilogie.fr/O/D'o%C3%B9_vient_la_L%C3%A9gion_d'Honneur_?"

I would argue that this is not a bug in linkchecker, however: that URL really *does* redirect to itself. The only reason it doesn't fail in curl is because curl doesn't follow redirects by default, while linkchecker does.

You can reproduce the problem with curl using the `-L` flag:

    $ curl -L "https://omnilogie.fr/O/D'o%C3%B9_vient_la_L%C3%A9gion_d'Honneur_?"
    curl: (47) Maximum (50) redirects followed

This is, therefore, not a bug and linkchecker is behaving as expected when flagging this URL as a problem. Why this would work in a browser is beyond my understanding however - it is possible something fishy is going on on the server side of things...

**Comment by Neamar** [comment]
Hi @anarcat!

You are correct for the apostrophe.

However, the curl call you gave works for me:

```
└[/tmp]> curl -L "https://omnilogie.fr/O/D'o%C3%B9_vient_la_L%C3%A9gion_d'Honneur_?" | head -n 15
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0<!DOCTYPE html>
<html>
<head>
        <meta charset="iso-8859-1" />
        <link rel="alternate" type="application/rss+xml" title="Flux RSS des articles" href="/flux.rss" />
        <link rel="stylesheet" media="all" href="/CSS/omni.responsive.min.css?v5508" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />

        <title>D'o vient la Lion d'Honneur&nbsp;?  | Un article d'Omnilogie.fr</title>
        <!--[if IE]>
                <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
        <link rel="canonical" href="https://omnilogie.fr/O/D'o_vient_la_Lion_d'Honneur_?" />
        <script type="text/javascript">
        var inits=[];
```

You mention that the URL redirects to itself, but that's also not something that I can reproduce using Chrome, where I do get a standard 200:

![image](https://user-images.githubusercontent.com/536844/51149016-306ab900-1858-11e9-8cb6-e154e53fc008.png)

For full disclosure: this is a website I created in 2010. It is ugly and old, and the URL convention is definitely super weird, but there isn't anything on the server side acting differently when the user agent is a browser.

You might be right and linkchecker is probably not to blame, however I disagree when you say that curl work because it doesn't follow redirect: even the initial curl retrieves the page HTML for me.
That's why I thought it could be something within linkchecker

**Comment by anarcat** [comment]
okay, that's super weird. i definitely can't make the site work with curl, so I don't know what's going on there.

```
anarcat@curie:dateparser(master)$ curl -L "https://omnilogie.fr/O/D'o%C3%B9_vient_la_L%C3%A9gion_d'Honneur_?"
curl: (47) Maximum (50) redirects followed
[47]anarcat@curie:dateparser(master)$ curl -V
curl 7.62.0 (x86_64-pc-linux-gnu) libcurl/7.62.0 OpenSSL/1.1.1a zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.35.1 librtmp/2.3
Release-Date: 2018-10-31
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
```

i do agree things seem to work in chrome, so no idea what's going on. i will reopen as an act of good faith though. ;)

**Comment by Neamar** [comment]
Thanks for reopening!

I was actually able to reproduce with another URL, and there is indeed something in linkchecker that prevents checking URL ending with a question mark:

See for instance this URL, which doesn't include any special character:

```
linkchecker "https://omnilogie.fr/O/Des_bonbons_ou_un_sort_?"
```

Real URL is displayed as `https://omnilogie.fr/O/Des_bonbons_ou_un_sort_` (notice the missing question mark).
This URL doesn't have any special characters in it, so the question mark is the issue here.

And this URL, which does include special characters, works:

`linkchecker "https://omnilogie.fr/O/Les_mots_crois%C3%A9s" -r 0`

**tl;dr: URL ending with a question mark are transformed by linkchecker, which removes the question mark automatically**

**Comment by cjmayo** [comment]
A `?` is a terminating character for a path: 
https://tools.ietf.org/html/rfc3986#section-3.3

If desired it should be percent encoded to `%3F`. Some quick testing suggests that works in LinkChecker.

So I guess a ? would ideally raise a syntax error - browsers may implement quirks but LinkChecker syntax checks every URL. But that's hard to implement because the Python standard library function that we use does this:

```
>>> urllib.parse.urlsplit("http://localhost/path_?#frag")
SplitResult(scheme='http', netloc='localhost', path='/path_', query='', fragment='frag')
```
